### PR TITLE
FILTER clause for aggregate functions

### PR DESCRIPTION
**Pull request by nickitat**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Issue #26776

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
It's better not to throw exception directly from parser, it's not expected.
Maybe just add second `If` modifier as usual and then handle it in AggregateFunctionFactory?

File: src/Parsers/ExpressionElementParsers.cpp [hunk 1]
```diff
@@ -504,6 +516,51 @@ bool ParserTableFunctionView::parseImpl(Pos & pos, ASTPtr & node, Expected & exp
     return true;
 }
 
+bool ParserFilterClause::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
+{
+    assert(node);
+    ASTFunction & function = dynamic_cast<ASTFunction &>(*node);
+
```
Comment (vdimir) [inline src/Parsers/ExpressionElementParsers.cpp:523]: Maybe lets make argument of type `ASTFunction` to be explicit for caller and have less change of passing wrong input into `parseImpl`

File: src/Parsers/ExpressionElementParsers.cpp [hunk 2]
```diff
@@ -504,6 +516,51 @@ bool ParserTableFunctionView::parseImpl(Pos & pos, ASTPtr & node, Expected & exp
...
+        throw Exception(
+            ErrorCodes::ILLEGAL_AGGREGATION,
+            "Filter clause provided for an aggregating function (" + function.name + ") already containing If suffix");
+    }
+    if (condition->children.empty())
+    {
+        throw Exception(ErrorCodes::SYNTAX_ERROR, "Empty condition for WHERE");
+    }
+
+    function.name += "If";
```
Comment (vdimir) [inline src/Parsers/ExpressionElementParsers.cpp:559]: Is to possible to have not aggregate function here, that doesn't support `If` modifier?

File: tests/queries/0_stateless/02000_select_with_filter.sql [hunk 3]
```diff
@@ -0,0 +1,4 @@
+SELECT argMax(number, number + 1) FILTER(WHERE number != 99) FROM numbers(100) ;
+SELECT sum(number) FILTER(WHERE number % 2 == 0) FROM numbers(100);
+SELECT sumIfOrNull(number, number % 2 == 1) FILTER(WHERE number % 2 == 0) FROM numbers(100); -- { clientError 184 }
+SELECT sum(number) FILTER(WHERE) FROM numbers(100); -- { clientError 62 }
```
Comment (vdimir) [inline tests/queries/0_stateless/02000_select_with_filter.sql:4]: There are some issues with AST fuzzer if we have unparsable queries in tests. But Actually first one have correct syntax, but we mustn't throw exception from parser

File: src/Parsers/ExpressionElementParsers.cpp (hunk 2 above)
Comment (vdimir) [reply src/Parsers/ExpressionElementParsers.cpp:559]: It should be handled, maybe not in parser but on later stages
```
:) SELECT identity(number) FILTER(WHERE number > 22) FROM numbers(100) ;

SELECT identityIf(number, number > 22)
FROM numbers(100)

Query id: 7001a006-0879-4206-a151-0ae842500997


0 rows in set. Elapsed: 0.017 sec.

Received exception from server (version 21.9.1):
Code: 46. DB::Exception: Received from localhost:9000. DB::Exception: Unknown function identityIf. Maybe you meant: ['identity']: While processing identityIf(number, number > 22). (UNKNOWN_FUNCTION)

------ 

:) SELECT throw(number == 1) FILTER(WHERE number > 22) FROM numbers(100) ;

SELECT throwIf(number = 1, number > 22)
FROM numbers(100)

Query id: 8c13959e-8b82-40a8-b155-516075b2de4f


0 rows in set. Elapsed: 0.003 sec.

Received exception from server (version 21.9.1):
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: Illegal type UInt8 of argument of function throwIf: While processing throwIf(number = 1, number > 22). (ILLEGAL_TYPE_OF_ARGUMENT)
```

File: src/Parsers/ExpressionElementParsers.cpp (hunk 1 above)
Comment (nickitat) [reply src/Parsers/ExpressionElementParsers.cpp:523]: `parseImpl` is declared in `IParserBase` and is called by `IParserBase::parse`, so we hardly can change its signature as long is it remains to be `IParserBase`'s descendant.

**Comment by nickitat** [comment]
now I throw in this [line](https://github.com/ClickHouse/ClickHouse/blob/master/src/AggregateFunctions/AggregateFunctionFactory.cpp#L173) for aggregates. but it breaks [this test](https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/00545_weird_aggregate_functions.sql).
I should try again )

File: src/Parsers/ExpressionElementParsers.cpp (hunk 2 above)
Comment (nickitat) [reply src/Parsers/ExpressionElementParsers.cpp:547]: > It should be handled, maybe not in parser but on later stages

I would say that CH should not happily accept the wrong query and produce the result (which was the case with sumIf + filter) but +- any error is enough for such weird cases (fooIf doesn't exist or resulting set of args is wrong).

**Comment by nickitat** [comment]
@vdimir take a look, pls

**Review by vdimir — APPROVED** [review]
Good work!

**Pull request merged**